src/uu/tac/src/tac.rs

```diff
@@ -82,96 +86,62 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
-    let mut exit_code = 0;
+fn generic_tac<T>(f: &mut T, separator: &str, before: bool) -> io::Result<()>
+where
+    T: Read + Seek,
+{
     let mut out = stdout();
-    let sbytes = separator.as_bytes();
-    let slen = sbytes.len();
+    let sep = separator.as_bytes().first().unwrap();
+    if before {
+        for line in rlines_leading_separator(f, *sep) {
+            out.write_all(&line)?;
+        }
+    } else {
+        for line in rlines_trailing_separator(f, *sep) {
+            out.write_all(&line)?;
+        }
+    }
+    Ok(())
+}
+
+/// Print lines of `stdin` in reverse.
+fn stdin_tac(separator: &str, before: bool) -> io::Result<()> {
+    let mut data = Vec::new();
+    stdin().read_to_end(&mut data).unwrap();
+    let mut file = Cursor::new(&data);
+    generic_tac(&mut file, separator, before)
+}
 
+/// Print lines of the given file in reverse.
+fn file_tac(filename: &str, separator: &str, before: bool) -> io::Result<()> {
+    let mut file = File::open(Path::new(filename))?;
+    generic_tac(&mut file, separator, before)
+}
+
+fn tac(filenames: Vec<String>, before: bool, _: bool, separator: &str) -> i32 {
+    let mut exit_code = 0;
     for filename in &filenames {
-        let mut file = BufReader::new(if filename == "-" {
-            Box::new(stdin()) as Box<dyn Read>
+        if filename == "-" {
+            if let Err(e) = stdin_tac(separator, before) {
+                show_error!("failed to read '{}': {}", filename, e);
+                exit_code = 1;
+            }
         } else {
-            let path = Path::new(filename);
-            if path.is_dir() || path.metadata().is_err() {
-                if path.is_dir() {
-                    show_error!("dir: read error: Invalid argument");
-                } else {
-                    show_error!(
-                        "failed to open '{}' for reading: No such file or directory",
-                        filename
-                    );
-                }
+            if Path::new(filename).is_dir() {
+                show_error!("{}: read error: Invalid argument", filename);
                 exit_code = 1;
                 continue;
             }
-            match File::open(path) {
-                Ok(f) => Box::new(f) as Box<dyn Read>,
-                Err(e) => {
-                    show_error!("failed to open '{}' for reading: {}", filename, e);
-                    exit_code = 1;
-                    continue;
+            if let Err(e) = file_tac(filename, separator, before) {
+                match e.kind() {
+                    ErrorKind::NotFound => {
+                        show_error!("failed to open '{}' for reading: {}", filename, e)
+                    }
+                    _ => show_error!("failed to read '{}': {}", filename, e),
                 }
+                exit_code = 1;
             }
-        });
-
-        let mut data = Vec::new();
-        if let Err(e) = file.read_to_end(&mut data) {
```

Comment:
Previously, the code was reading the entire file into memory. This pull request replaces that with reading a file by chunks in reverse. (But for `stdin` we still read the entire stream into memory.)